[SE-0456] Span properties

glessard opened the pull request:
Implements most of the `span` properties proposed [here](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0456-stdlib-span-properties.md), reviewed [here](https://forums.swift.org/t/77233), and accepted [here](https://forums.swift.org/t/77684).

Addresses most of rdar://137710901

`Span`s vended from `InlineArray` and `CollectionOfOne` are disabled at the moment.

`span` properties for `String.UTF8View` and `Substring.UTF8View` are added [here](https://github.com/swiftlang/swift/pull/80116).

glessard committed:
[stdlib] add `bytes` property to `Span`

glessard committed:
[test] test `Span.bytes`

glessard committed:
[test] span properties for `Array` family

glessard committed:
[stdlib] refactor lazy-eager-bridging for reusability

glessard commented:
@swift-ci please build toolchain

DougGregor commented on stdlib/public/core/CollectionOfOne.swift:
```diff
@@ -22,6 +22,7 @@
 ///     let b = a + CollectionOfOne(toAdd)
 ///     // b == [1, 2, 3, 4, 100]
 @frozen // trivial-implementation
+@_addressableForDependencies
```
Does this have ABI implications?

glessard replied:
It does, specifically for the new case of functions returning non-escapable values. For existing functions returning escapable values it changes nothing. @jckarter knows more.

DougGregor commented on stdlib/public/core/Span/Properties/ArrayTypeExtensions.swift:
```diff
@@ -0,0 +1,59 @@
+
+extension ContiguousArray {
+
+  @available(SwiftStdlib 6.1, *)
```
Should this be `SwiftStdlib 6.2`?

glessard commented:
@swift-ci please build toolchain

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please smoke test macOS platform

glessard commented:
@swift-ci please build toolchain

glessard commented:
@swift-ci please test linux platform

DougGregor commented:
@swift-ci please build toolchain Windows

glessard commented:
@swift-ci please build toolchain

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please test macOS platform

glessard commented:
@swift-ci please build toolchain

benrimmington commented on test/stdlib/Span/StringUTF8StorageProperty.swift:
```diff
@@ -0,0 +1,105 @@
+.code {
+  guard #available(SwiftStdlib 6.1, *) else { return }
+
+  let s = "A small string.".utf8
```
This will be a large string on platforms where [`_SmallString.capacity`](https://github.com/swiftlang/swift/blob/swift-6.0.3-RELEASE/stdlib/public/core/SmallString.swift#L79-L91) is 10 or 14. Also, does it matter that small strings don't store a [zero-terminator](https://github.com/swiftlang/swift/blob/swift-6.0.3-RELEASE/stdlib/public/core/SmallString.swift#L219-L225)?

The new properties are `@available(SwiftStdlib 6.2, *)`, but these new test suites are checking for `SwiftStdlib 6.1` or later.

glessard replied:
`Span<UInt8>` won't care about a zero terminator. C-string interop will be more specifically handled by the pitched `UTF8Span`. The availability checking in the tests will be fixed; thanks for the reminder.

jckarter commented:
@swift-ci Please test

benrimmington commented on stdlib/public/core/Span/Properties/InlineTypeExtensions.swift:
```diff
@@ -0,0 +1,123 @@
+  }
+}
+
+extension SIMD2 {
+
+  @available(SwiftStdlib 6.2, *)
+  public var storage: Span<Scalar> {
+    @_alwaysEmitIntoClient
+    @lifetime(borrow self)
+    borrowing get {
+      let pointer = UnsafePointer<Scalar>(Builtin.addressOfBorrow(self))
+      let span = Span(_unsafeStart: pointer, count: 2)
+      return _overrideLifetime(span, borrowing: self)
+    }
+  }
+}
```
The generic SIMD types don't require *inline* or contiguous storage. For example, [SIMD-nontrivial.swift](https://github.com/swiftlang/swift/blob/swift-6.0.3-RELEASE/test/stdlib/SIMD-nontrivial.swift) uses an `Array<Scalar>`. Could these properties be moved into the existing [SIMDConcreteOperations.swift.gyb](https://github.com/swiftlang/swift/blob/swift-6.0.3-RELEASE/stdlib/public/core/SIMDConcreteOperations.swift.gyb) extensions? (Where the storage is known to be either `Builtin.Vec…xInt…` or `Builtin.Vec…xFPIEEE…`)

glessard replied:
Sorry for missing this earlier. You are correct that these would have to be defined specifically on the concrete types; it isn't clear whether that is worth the code size, so we will defer the SIMD-specific extensions until later. We may define them on the concrete type, or we may fix the `SIMD` protocol, or we may take advantage of `InlineArray` instead.

benrimmington replied:
> … we will defer the SIMD-specific extensions until later.

Another issue was that your extensions in this file would have needed either:

* [`#if SWIFT_STDLIB_ENABLE_VECTOR_TYPES`](https://github.com/swiftlang/swift/blob/swift-6.0.3-RELEASE/stdlib/public/core/StringUTF16View.swift#L725) in Swift, or

* [`if(SWIFT_STDLIB_ENABLE_VECTOR_TYPES)`](https://github.com/swiftlang/swift/blob/swift-6.0.3-RELEASE/stdlib/public/core/CMakeLists.txt#L257) in CMake.

natecook1000 commented:
@swift-ci Please build toolchain macOS platform

natecook1000 commented:
@swift-ci Please build toolchain macOS platform

glessard committed:
[stdlib] add `span` properties to UnsafeBufferPointer family

glessard committed:
[stdlib] add `span` properties to array types

glessard committed:
[stdlib] add `span` property to `CollectionOfOne`

glessard committed:
[stdlib] add `span` property to `InlineArray`

glessard committed:
[test] ritual acts of contrition

benrimmington commented on stdlib/public/core/Span/RawSpan.swift:
```diff
@@ -286,18 +286,19 @@ extension RawSpan {
+    let pointer = span._pointer
+    let rawSpan = RawSpan(
+      _unchecked: pointer,
+      byteCount: span.count == 1 ? MemoryLayout<Element>.size
+                 : span.count &* MemoryLayout<Element>.stride
```
If this is correct then other APIs may need to be updated:

* `RawSpan.init(_unsafeStart:count:)` which uses `count * MemoryLayout<T>.stride`.

* `Span.withUnsafeBytes(_:)` which uses `_count * MemoryLayout<Element>.stride`.

* `Span.indices(of:)` which uses `stride &* other._count` and `stride &* _count`.

* `Span.init(_bytes:)` that consumes a `RawSpan`.

Alternatively, perhaps `CollectionOfOne` should be removed from the proposal?

benrimmington replied:
> The idea here is that safely converting a `Span` of 1 probably doesn't want to use `stride`; the need indeed comes from `CollectionOfOne`.

The same issue could affect nonempty `Span`s of any length:

```swift
let one: StaticString = ""
let two: (StaticString, StaticString) = ("", "")

withUnsafePointer(to: one) { start in
  let buffer = UnsafeBufferPointer(start: start, count: 1)
  return buffer.span.bytes.byteCount //-> either 17 or 24.
}

withUnsafePointer(to: two) { start in
  let start = start.pointer(to: \.0)
  let buffer = UnsafeBufferPointer(start: start, count: 2)
  return buffer.span.bytes.byteCount //-> either 41 or 48.
}
```

Should the padding after the *last* element always be excluded?

glessard replied:
Using tuples in this way is incorrect and always has been, because of that padding issue. The buffer in the `wUP(to: two)` closure is already invalid. We do get away with it with C imports because in that case `stride` is defined to be equal to `size`. We would really prefer to get this pattern (buffer-to-tuple) moved over to `InlineArray`.

benrimmington commented:
Will `KeyValuePairs` be a separate pull request?

Some of the test filenames and `TestSuite` names contain "Storage" instead of "Span".

The api-digester tests may need:

```
Struct InlineArray is now with @_addressableForDependencies
```

glessard commented on stdlib/public/core/CollectionOfOne.swift:
```diff
@@ -158,6 +159,20 @@ extension CollectionOfOne: RandomAccessCollection, MutableCollection {
+  @available(SwiftStdlib 6.2, *)
+  public var span: Span<Element> {
+    @lifetime(borrow self)
+    @_alwaysEmitIntoClient
```
Not sure! I will find out.

glessard replied:
`@_addressableSelf` shouldn't be needed for `@_addressableForDependencies` types.
(I removed it from `InlineArray`'s version.)

glessard commented:
> Will `KeyValuePairs` be a separate pull request?

I forgot to include it!

> Some of the test filenames and `TestSuite` names contain "Storage" instead of "Span".

Will review those.

> The api-digester tests may need:
> 
> ```
> Struct InlineArray is now with @_addressableForDependencies
> ```

My local test run didn't need it; the annotation might be redundant because of the `Builtin.FixedArray` field. Will inquire.

glessard committed:
[test] better names for test suites

glessard committed:
[stdlib] add `span` property to `KeyValuePairs`

glessard committed:
[test] rename files

glessard committed:
[stdlib] remove an unneeded annotation

glessard committed:
[stdlib] only enable “AddressableTypes” for the core stdlib

glessard commented:
@swift-ci please test Windows platform

glessard commented:
@swift-ci please test Linux platform

glessard commented:
@swift-ci please smoke test macOS platform

etcwilde reviewed (approved):
Approving changes to the new runtime build

etcwilde commented on Runtimes/Core/CMakeLists.txt:
```diff
@@ -193,6 +193,7 @@ add_compile_options(
   "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature Extern>"
   "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature ValueGenerics>"
   "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature AddressableParameters>"
+  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature AddressableTypes>"
```
Is this needed on all Swift libraries (concurrency, swiftCore, and SwiftOnoneSupport) or just libswiftCore?

glessard replied:
For now it's just libswiftCore.

etcwilde replied:
We could put the flag just on the `swiftCore` target in `core/` to make it clear that it's the only target that needs it, but since it's an experimental-feature flag that will eventually be applied everywhere once the compiler enables it fully, I am not going to block the patch on it.

glessard replied:
I asked @jckarter what he'd prefer to do here, and I'll move AddressableParameters and AddressableTypes to the libswiftCore module config.

glessard commented:
@swift-ci please test Linux platform

glessard commented:
@swift-ci please test Windows platform

benrimmington commented on stdlib/public/core/KeyValuePairs.swift:
```diff
@@ -125,6 +125,19 @@ extension KeyValuePairs: RandomAccessCollection {
+extension KeyValuePairs {
+  
+  @available(SwiftStdlib 6.2, *)
+  public var span: Span<(Key, Value)> {
```
Can this convert to the tuple with labels?

```suggestion
  public var span: Span<Element> {
```

glessard replied:
Done, though it's a slightly more hair-raising changing than just spelling it `Element`.

benrimmington commented on test/stdlib/Span/ArraySpanProperties.swift:
```diff
@@ -0,0 +1,92 @@
+//===--- ArrayStorage.swift ---- test span properties for Array types -----===//
```
```suggestion
//===--- ArraySpanProperties.swift ----------------------------------------===//
```

benrimmington commented on test/stdlib/Span/InlineSpanProperties.swift:
```diff
@@ -0,0 +1,88 @@
+//===--- InlineStorage.swift ----------------------------------------------===//
```
```suggestion
//===--- InlineSpanProperties.swift ---------------------------------------===//
```

glessard committed:
[stdlib] match `KeyValuePairs`’s `Element` type

glessard committed:
Update test/stdlib/Span/ArraySpanProperties.swift

Co-authored-by: Ben Rimmington <me@benrimmington.com>

glessard committed:
Update test/stdlib/Span/InlineSpanProperties.swift

Co-authored-by: Ben Rimmington <me@benrimmington.com>

natecook1000 commented:
@swift-ci Please build toolchain macOS platform

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please smoke test

glessard committed:
[test] add a test case

glessard commented:
@swift-ci please build toolchain macOS platform

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please build toolchain

glessard committed:
[stdlib] Update availability annotations for `Span`

glessard committed:
[stdlib] add newly-required unsafe annotations

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please build toolchain

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please build toolchain

edymtt commented on Runtimes/Core/CMakeLists.txt:
```diff
@@ -191,7 +191,6 @@ add_compile_options(
   "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature BitwiseCopyable>"
   "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature Extern>"
   "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature ValueGenerics>"
-  "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-enable-experimental-feature AddressableParameters>"
```
I believe this may be a diff against an old commit, since we moved these flags in #79677 -- you may want to consider removing this from `Runtimes/Core/cmake/modules/ExperimentalFeatures.cmake` instead

glessard replied:
Will do, thanks.

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please build toolchain

glessard committed:
[gardening] update some copyright years

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please test

glessard committed:
[temporary] disable spans over inline elements

glessard commented:
@swit-ci please test

glessard committed:
[stdlib] changes to lifetime annotations

Changes to lifetime annotations implemented in https://github.com/swiftlang/swift/pull/80064

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please test linux platform

glessard commented:
@swift-ci please test macOS platform

glessard commented:
@swift-ci please test macOS platform

glessard commented:
@swift-ci please test linux platform

glessard committed:
[test] amend test command

glessard commented:
@swift-ci please smoke test

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please test

glessard commented:
@swift-ci please test windows platform
